File: q40-q49/q46/q46.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> solution;
        helper(solution, nums, 0);
        return solution;
    }
    void helper(vector<vector<int>>& solution, vector<int>& nums, int current_index){
        if (current_index == nums.size()){
            solution.push_back(nums);
            return;
        }
        for (int i = current_index; i < nums.size(); i++){
            int temp = nums[i];
            nums[i] = nums[current_index];
            nums[current_index] = temp;
            helper(solution, nums, current_index + 1);
            temp = nums[i];
            nums[i] = nums[current_index];
            nums[current_index] = temp;
        }
    }
};

File: q40-q49/q46/q46.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans = []
        self.helper(nums, [], ans)
        return ans

    def helper(self, nums, temp, ans):
        if len(nums)==0:
            ans.append(temp)
            return

        for i in range(len(nums)):
            self.helper(nums[:i]+nums[i+1:],temp+[nums[i]],ans )
    
            
